src/ipa_pc/mod.rs

```diff
@@ -1204,146 +1223,214 @@ impl<G: AffineCurve, D: Digest> PolynomialCommitment<G::ScalarField> for InnerPr
+                G::ScalarField::one()
+            ]);
+
+            // z(X) = PROD (X - x_i)
+            z_polynomial = &z_polynomial * &x_polynomial;
 
             // (p_i(X) - y_i) / (X - x_i)
             let polynomial =
-                &(labeled_polynomial.polynomial() - &Polynomial::from_coefficients_vec(vec![evaluated_y]))
+                &(labeled_polynomial.polynomial() - &Polynomial::from_coefficients_slice(&[y_i]))
                     /
-                    &Polynomial::from_coefficients_vec(vec![
-                        (G::ScalarField::zero() - point),
-                        G::ScalarField::one()
-                    ]);
+                    &x_polynomial;
 
             // h(X) = SUM( lambda^i * ((p_i(X) - y_i) / (X - x_i)) )
-            batch_polynomial += (cur_challenge, &polynomial);
+            h_polynomial += (cur_challenge, &polynomial);
 
             // lambda^i
             cur_challenge = cur_challenge * &lambda;
         }
+
         end_timer!(h_poly_time);
 
         let key_len = ck.comm_key.len();
-        let p_len = batch_polynomial.coeffs.len();
-        let segments_count = std::cmp::max(1, p_len / key_len + if p_len % key_len != 0 { 1 } else { 0 });
-
-        let batch_randomness = if has_hiding {
-            if rng.is_none() {
-                Err(Error::Other("Rng not set".to_owned()))?
-            }
-            Randomness::rand(segments_count, false, rng.as_mut().unwrap())
-        } else {
-            Randomness::empty(segments_count)
-        };
 
-        // Commitment of the h(X) polynomial
-        let commit_time = start_timer!(|| format!("Commit to batch polynomial of degree {}", batch_polynomial.degree()));
-        let batch_commitment: Vec<G>;
+        let h_polynomial_len = h_polynomial.coeffs.len();
+        let lc_polynomial_len = h_polynomial_len + 1;
+        let h_segments_count = std::cmp::max(1, h_polynomial_len / key_len + if h_polynomial_len % key_len != 0 { 1 } else { 0 });
+        let lc_segments_count = std::cmp::max(1, lc_polynomial_len / key_len + if lc_polynomial_len % key_len != 0 { 1 } else { 0 });
 
-        if p_len > key_len {
+        let commit_time = start_timer!(|| format!("Commit to h(X) polynomial of degree {}", h_polynomial.degree()));
 
-            batch_commitment = (0..segments_count).into_iter().map(
+        // Commitment to the h(X) polynomial
+        let h_comm = if h_polynomial_len > key_len {
+            (0..h_segments_count).into_iter().map(
                 |i| {
                     Ok(Self::cm_commit(
                         &ck.comm_key,
-                        &batch_polynomial.coeffs[i * key_len..core::cmp::min((i + 1) * key_len, p_len)],
-                        Some(ck.s),
-                        Some(batch_randomness.rand[i]),
+                        &h_polynomial.coeffs[i * key_len..core::cmp::min((i + 1) * key_len, h_polynomial_len)],
+                        None,
+                        None,
```

Comment:
h(X) must be committed using hiding randomnesses, if one of the involved polynomials has hiding=true.

Reply:
Done